src/tiled/changeworld.cpp

```diff
@@ -29,74 +29,90 @@
 
 namespace Tiled {
 
-AddMapCommand::AddMapCommand(const QString &worldName, const QString &mapName, const QRect &rect)
-    : QUndoCommand(QCoreApplication::translate("Undo Commands", "Add Map to World"))
-    , mWorldName(worldName)
+AddRemoveMapCommand::AddRemoveMapCommand(WorldDocument *worldDocument,
+                                         const QString &mapName,
+                                         const QRect &rect,
+                                         QUndoCommand *parent)
+    : QUndoCommand(parent)
+    , mWorldDocument(worldDocument)
     , mMapName(mapName)
     , mRect(rect)
 {
 }
 
-void AddMapCommand::undo()
+void AddRemoveMapCommand::addMap()
 {
-    WorldManager::instance().removeMap(mMapName);
+    auto world = mWorldDocument->world();
+    world->addMap(mMapName, mRect);
+    emit mWorldDocument->worldChanged();
 }
 
-void AddMapCommand::redo()
+void AddRemoveMapCommand::removeMap()
 {
-    WorldManager::instance().addMap(mWorldName, mMapName, mRect);
+    auto world = mWorldDocument->world();
+
+    const int index = world->mapIndex(mMapName);
+    if (index < 0)
+        return;
+
+    world->removeMap(index);
+    emit mWorldDocument->worldChanged();
```

Comment:
Each change affecting a world just emits `worldChanged` now, whereas before the `WorldManager::worldsChanged` signal was emitted. Probably we should forward the former to the latter, because existing code relies on getting `worldsChanged` (and often, we need to respond to any world changing, because we do not know beforehand which world a map will then belong to).

Reply:
It appears I already added the connection between `worldChanged` and `worldsChanged`.